src/components/TextField/TextField.js

```diff
@@ -0,0 +1,317 @@
+                        >
+                            {tooltip}
+                        </TooltipBox>
+                    }
+                >
+                    {input}
+                </Tooltip> :
+                input
+        );
+    }
+
+    render() {
+        /* eslint-disable jsx-a11y/label-has-for */
+        // @NB: jsx-a11y/label-has-for fails with UID as id
+        const showLabel = (
+            this.props.hasLabelAlways ||
+            this.state.hasFocus ||
+            this.state.hasMouseOver ||
+            !this.state.value
+        );
+        const requiredIcon = this.props.isRequired ?
+            <Tooltip tooltip={this.props.tooltipRequired}><IconRequired /></Tooltip> :
+            null;
+        const label = this.props.label && showLabel ?
+            (
+                <label
+                    htmlFor={this.uid}
+                    className={cx(
+                        'uir-textfield-label',
+                        {
+                            'uir-textfield-label--visually-hidden': this.props.variant === TextFieldVariant.TITLE,
+                        },
+                    )}
+                >
+                    {this.props.label}
+                    {requiredIcon}
+                </label>
+            ) :
+            null;
+        const showPlaceholder = (
+            !this.props.label ||
+            this.state.hasFocus ||
+            this.props.variant === TextFieldVariant.TITLE
+        );
+        const icon = this.props.icon ?
+            this.props.icon :
+            null;
+        const clearIcon = this.props.isClearable && this.state.value ?
+            (
+                <Button
+                    onClick={this.handleClearIconClick}
+                    variant={ButtonVariant.CLEAR}
+                >
+                    <IconClear />
+                </Button>
+            ) :
+            null;
+        return (
+            <div
+                className={cx(
+                    'uir-textfield',
+                    {
+                        'uir-textfield--disabled': this.props.isDisabled,
+                        'uir-textfield--focus': this.state.hasFocus,
+                        'uir-textfield--full-width': this.props.isFullWidth,
+                        'uir-textfield--has-left-icon': this.props.icon,
+                        'uir-textfield--has-right-icon': this.props.isRequired || this.props.isClearable,
+                        'uir-textfield--has-value': this.state.value,
+                        'uir-textfield--invalid': this.props.isValid === false,
+                        'uir-textfield--readonly': this.props.isReadOnly,
+                        'uir-textfield--title': this.props.variant === TextFieldVariant.TITLE,
+                        'uir-textfield--valid': this.props.isValid,
+                    },
+                    this.props.className,
+                )}
+                id={this.props.id}
```

Comment:
I'm not sure if I agree with this. When providing a TextField `id`, I would expect it to be applied to the actual field - the input element.
Also, this reminds me: we need a `name` prop too to control the `name` attribute of the input.

Reply:
Question is, why do we have the `id` prop?

If it's to allow the user to identify (and then manipulate) the component, then an `id` on the wrapper does work.  If the user needs the input they can do the equivilant of `$('#compid  input')`

If on the other hand, the `id` is to identify the input then perhaps `id` should be linked to the input, not the wrapper, however, we give the `input` an automatically assigned `id` to allow the label to be linked via `for`

On balance I think I would leave it as it is ;)